lib/basedriver/capabilities.js

```diff
@@ -0,0 +1,111 @@
+
+// Takes primary caps object and merges it into a secondary caps object.
+// (see https://www.w3.org/TR/webdriver/#dfn-merging-capabilities)
+function mergeCaps (primary = {}, secondary = {}) {
+  let result = Object.assign(primary);
+
+  for (let [name, value] of _.toPairs(secondary)) {
+    // Overwriting is not allowed. Primary and secondary must have different properties (w3 rule 4.4)
+    if (!_.isUndefined(primary[name])) {
+      throw new Error(`property ${name} should not exist on both primary and secondary`);
+    }
+    result[name] = value;
+  }
+
+  return result;
+}
+
+// Validates caps against a set of constraints
+function validateCaps (caps, constraints = {}, skipPresenceConstraint = false) {
+  let message = '';
+  if (!_.isObject(caps)) {
+    throw new Error(`must be a JSON object`);
+  }
+
+  constraints = _.cloneDeep(constraints); // Defensive copy
+
+  // Remove the 'presence' constraint if we're not checking for it
+  if (skipPresenceConstraint) {
+    constraints = Object.assign(constraints);
+    for (let key of Object.keys(constraints)) {
+      delete constraints[key].presence;
+    }
+  }
+
+  let validationErrors = validator.validate(_.pickBy(caps, util.hasValue),
+                                              constraints,
+                                              {fullMessages: false});
+                                          
+  if (validationErrors) {
+    for (let [attribute, reasons] of _.toPairs(validationErrors)) {
+      for (let reason of reasons) {
+        message += ` ${attribute} ${reason},`;
+      }
+    }
+    message = `${message.slice(0, -1)}.`;
+    throw new Error(message);
+  }
+
+  // Return caps
+  return caps;
+}
+
+// Parse capabilities (based on https://www.w3.org/TR/webdriver/#processing-capabilities)
+function parseCaps (caps, constraints = {}, shouldValidateCaps = true) {
+  if (!_.isObject(caps)) {
+    throw new Error('The capabilities argument was not valid for the following reason(s): "capabilities" must be a JSON object.');
+  }
+
+  let {alwaysMatch: requiredCaps, firstMatch: allFirstMatchCaps = []} = caps; // jshint ignore:line
+
+  if (!_.isObject(requiredCaps)) {
+    requiredCaps = {}; 
+  }
+
+  requiredCaps = validateCaps(requiredCaps, constraints);
+
+  // Reject 'firstMatch' argument if it's not an array (see spec #3.2)
+  if (!_.isArray(allFirstMatchCaps)) {
+    throw new Error('The capabilities.firstMatch argument was not valid for the following reason(s): "capabilities.firstMatch" must be a JSON array or undefined');
+  }
+
+  // Validate the firstMatch capabilities one-by one (see spec #5)
+  let validatedFirstMatchCaps = allFirstMatchCaps.map((firstMatchCaps, index) => {
+    try {
+      return shouldValidateCaps ? validateCaps(firstMatchCaps, constraints) : firstMatchCaps;
+    } catch (e) {
+      throw new Error(`The capabilities.firstMatch argument was not valid for the following reason: "capabilities.firstMatch[${index}] ${e.message}`);
```

Comment:
Does `capabilities.firstMatch[${index}]` have a value? Is it a key? Perhaps it should be logged? This just feels confusing as a log line.